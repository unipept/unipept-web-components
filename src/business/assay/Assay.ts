import { CountTable } from "@/business/counts/CountTable";
import Visitable from "@/logic/patterns/visitor/Visitable";

export default abstract class Assay implements Visitable<AssayVisitor> {
    protected constructor(
        public readonly changeListeners: ChangeListener<Assay>[],
        public readonly id: string,
        protected name?: string,
        protected date?: Date,
    ) {}

    public getId(): string {
        return this.id;
    }

    public getName() {
        return this.name;
    }

    public setName(name: string) {
        const oldName: string = this.name;
        this.name = name;
        this.onUpdate("name", oldName, name);
    }

    public getDate() {
        return this.date;
    }

    public setDate(date: Date) {
        const oldDate: Date = date;
        this.date = date;
        this.onUpdate("date", oldDate, date);
    }

    protected onUpdate(field: string, oldValue: any, newValue: any) {
        for (const changeListener of this.changeListeners) {
            changeListener.onChange(this, field, oldValue, newValue);
        }
    }

    abstract async accept(visitor: AssayVisitor): Promise<void>;
}

