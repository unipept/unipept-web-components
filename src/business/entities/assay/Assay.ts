import ChangeListener from "./../ChangeListener";
import AssayVisitor from "./AssayVisitor";
import Visitable from "./../../visitor/Visitable";

export default abstract class Assay implements Visitable<AssayVisitor> {
    private name: string = "";
    private date: Date = new Date();
    private changeListeners: ChangeListener<Assay>[] = [];

    protected constructor(
        public id: string
    ) {}

    public addChangeListener(listener: ChangeListener<Assay>): void {
        this.changeListeners.push(listener);
    }

    public getId(): string {
        return this.id;
    }

    public getName() {
        return this.name;
    }

    public setName(name: string) {
        const oldName: string = this.name;
        this.name = name;
        if (oldName !== name) {
            this.onUpdate("name", oldName, name);
        }
    }

    public getDate() {
        return this.date;
    }

    public setDate(date: Date) {
        const oldDate: Date = date;
        this.date = date;
        if (!oldDate || date.getTime() !== oldDate.getTime()) {
            this.onUpdate("date", oldDate, date);
        }
    }

    protected onUpdate(field: string, oldValue: any, newValue: any) {
        for (const changeListener of this.changeListeners) {
            changeListener.onChange(this, field, oldValue, newValue);
        }
    }

    abstract async accept(visitor: AssayVisitor): Promise<void>;
}

