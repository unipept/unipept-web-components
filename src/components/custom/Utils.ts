export default class Utils {
    /**
     * This method should be used when a specific URL should be opened in a new browser window. The method automatically
     * decides whether Electron or a default redirection should take place.
     * 
     * @param url The full url to which navigation should take place.
     */
    static openInBrowser(url: string): void {
        if (Utils.isElectron()) {
            const shell = require("electron").shell;
            shell.openExternal(url);
        } else {
            window.open(url);
        }
    }
    
    /**
     * Method that can be used to check if the app is being executed by Electron or something else.
     * 
     * @return True if the app is being executed in an Electron-powered environment.
     */
    static isElectron(): boolean {
        const userAgent = navigator.userAgent.toLowerCase();
        return userAgent.indexOf(" electron/") > -1;
    }
}
