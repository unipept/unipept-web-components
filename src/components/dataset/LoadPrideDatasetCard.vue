<docs>
The `LoadPrideDatasetCard` allows the user to load any dataset from the PRIDE-archive. The user enters a valid PRIDE id 
after which the required information is downloaded and transformed into an Assay.
</docs>

<template>
    <v-card flat>
        <v-card-text>
            <h3>Load data from the PRIDE archive</h3>
            <p>You can easily load data from the <a href="http://www.ebi.ac.uk/pride/" target="_blank">PRIDE</a> data repository. Simply enter an assay id (e.g. 8500) in the field below and click the 'Load PRIDE Dataset' button. The corresponding dataset will then be fetched using the PRIDE API and loaded into the search form on the left.</p>
            <v-form ref="prideAssayForm" @submit.prevent>
                <v-text-field label="Assay id" placeholder="e.g. 8500" :disabled="prideLoading" v-model="prideAssay" :rules="[value => !!value || 'Please enter a valid PRIDE assay number']" clearable></v-text-field>
            </v-form>
            <div class="card-actions">
                <v-btn v-if="!prideLoading" @click="fetchPrideAssay()">
                    <v-icon left>mdi-cloud-download</v-icon>
                    Fetch PRIDE dataset
                </v-btn>
                <v-progress-linear v-if="prideLoading" v-model="prideProgress"></v-progress-linear>
            </div>
            <dataset-form ref="prideDatasetForm" v-on:peptide-change="pridePeptides = $event" :peptides="pridePeptides" v-on:name-change="prideName = $event" :name="prideName" v-on:save-change="prideSave = $event" :save="prideSave" :loading="prideLoading"></dataset-form>
            <div class="card-actions">
                <v-btn :disabled="prideLoading" @click="selectPrideAssay()">
                    <v-icon left>mdi-plus</v-icon>
                    Add to selected datasets
                </v-btn>
            </div>
            <snackbar :timeout="0" ref="prideSnackbar">Loading dataset... <div class="spinner"></div></snackbar>
        </v-card-text>
    </v-card>
</template>

<script lang="ts">
import Vue from "vue";
import Component, { mixins } from "vue-class-component"
import DatasetMixin from "./DatasetMixin.vue";
import DatasetManager from "../../logic/data-management/DatasetManager";
import DatasetForm from "./DatasetForm.vue";
import Snackbar from "../custom/Snackbar.vue";

@Component({
    components: {
        Snackbar,
        DatasetForm
    }
})
export default class LoadPrideDatasetCard extends mixins(DatasetMixin) {
    $refs!: {
        prideAssayForm: any,
        prideDatasetForm: DatasetForm,
        prideSnackbar: Snackbar
    }

    private prideAssay: string = "";
    private pridePeptides: string = "";
    private prideName: string = "";
    private prideSave: boolean = true;
    private prideLoading: boolean = false;
    private prideProgress: number = 0;

    private fetchPrideAssay(): void {
        if (this.$refs.prideAssayForm.validate()) {
            this.prideLoading = true;
            let datasetManager: DatasetManager = new DatasetManager();
            let prideNumber: number = parseInt(this.prideAssay);

            this.prideName = "PRIDE assay " + prideNumber.toString();

            this.$refs.prideSnackbar.show();
            datasetManager
                .loadPrideDataset(prideNumber, (progress) => this.prideProgress = progress * 100)
                .then((peptides) => {
                    this.pridePeptides = peptides.join("\n");
                    this.prideLoading = false;
                    this.$refs.prideSnackbar.destroy();
                });
        }
    }

    private selectPrideAssay() {
        if (this.$refs.prideDatasetForm.isValid()) {
            this.storeDataset(this.pridePeptides, this.prideName, this.prideSave);
        }
    }
}
</script>

<style lang="less" scoped>
</style>
