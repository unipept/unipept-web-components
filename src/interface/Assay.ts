export default interface Assay {
    id: string
    name: string
    peptides: string[]
    amountOfPeptides: number
    createdAt: Date
};
