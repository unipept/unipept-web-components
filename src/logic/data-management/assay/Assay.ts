import DataRepository from "../../data-source/DataRepository";
import Visitable from "../../patterns/visitor/Visitable";
import AssayVisitor from "src/logic/data-management/assay/AssayVisitor";
import { StorageType } from "../StorageType";
import MPAConfig from "../MPAConfig";
import Entity from "./Entity";
import ChangeListener from "@/logic/data-management/ChangeListener";

export default abstract class Assay implements Visitable<AssayVisitor>, Entity<string> {
    protected id: string;
    protected name: string;
    protected date: Date;
    protected changeListener: ChangeListener;
    protected storageType: StorageType;

    protected _dataRepository: DataRepository;

    public progress: number = 0;

    constructor(changeListener?: ChangeListener, id?: string, storageType?: StorageType, name?: string, date?: Date) {
        this.id = id;
        this.name = name;
        this.date = date;
        this.changeListener = changeListener;
        this.storageType = storageType;
    }

    getId(): string {
        return this.id;
    }

    setId(id: string) {
        this.id = id;
        this.changeListener.onChange("id");
    }

    getName() {
        return this.name;
    }

    setName(name: string) {
        this.name = name;
        this.changeListener.onChange("name");
    }

    getDate() {
        return this.date;
    }

    setDate(date: Date) {
        this.date = date;
        this.changeListener.onChange("date");
    }

    getStorageType() {
        return this.storageType;
    }

    setStorageType(storageType: StorageType) {
        this.storageType = storageType;
    }

    get dataRepository() {
        return this._dataRepository;
    }

    getDateFormatted(): string {
        // @ts-ignore
        return this.date.getFullYear() + "/" + (this.date.getMonth() + 1).toString().padStart(2, "0") + "/" + this.date.getDate().toString().padStart(2, "0");
    }

    abstract async initDataRepository(mpaConfig: MPAConfig, baseUrl: string);
    abstract async accept(visitor: AssayVisitor): Promise<void>;
}
