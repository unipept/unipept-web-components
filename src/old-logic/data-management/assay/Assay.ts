import DataRepository from "../../data-source/DataRepository";
import Visitable from "../../patterns/visitor/Visitable";
import AssayVisitor from "./AssayVisitor";
import { StorageType } from "../StorageType";
import MPAConfig from "../MPAConfig";
import Entity from "./Entity";
import ChangeListener from "./../ChangeListener";

export default abstract class Assay implements Visitable<AssayVisitor>, Entity<string> {
    protected readonly id: string;
    protected name: string;
    protected date: Date;
    protected changeListeners: ChangeListener<Assay>[];
    protected storageType: StorageType;

    protected _dataRepository: DataRepository;

    public progress: number = 0;

    constructor(changeListeners: ChangeListener<Assay>[], id: string, storageType?: StorageType, name?: string, date?: Date) {
        this.id = id;
        this.name = name;
        this.date = date;
        this.changeListeners = changeListeners;
        this.storageType = storageType;
    }

    getId(): string {
        return this.id;
    }

    getName() {
        return this.name;
    }

    setName(name: string) {
        const oldName: string = this.name;
        this.name = name;
        this.onUpdate("name", oldName, name);
    }

    getDate() {
        return this.date;
    }

    setDate(date: Date) {
        const oldDate: Date = date;
        this.date = date;
        this.onUpdate("date", oldDate, date);
    }

    getStorageType() {
        return this.storageType;
    }

    setStorageType(storageType: StorageType) {
        this.storageType = storageType;
    }

    get dataRepository() {
        return this._dataRepository;
    }

    getDateFormatted(): string {
        // @ts-ignore
        return this.date.getFullYear() + "/" + (this.date.getMonth() + 1).toString().padStart(2, "0") + "/" + this.date.getDate().toString().padStart(2, "0");
    }

    public reset(): void {
        this.progress = 0;
        this._dataRepository = undefined;
    }

    protected onUpdate(field: string, oldValue: any, newValue: any) {
        for (const changeListener of this.changeListeners) {
            changeListener.onChange(this, field, oldValue, newValue);
        }
    }

    abstract async initDataRepository(mpaConfig: MPAConfig, baseUrl: string);
    abstract async accept(visitor: AssayVisitor): Promise<void>;
}
