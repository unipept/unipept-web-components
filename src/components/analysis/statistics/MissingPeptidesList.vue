<template>
    <div style="margin-top: 16px;">
        <div>
            Sorry, we didn't manage to find {{ missedPeptides.length }} of your
            peptides. You can BLAST them by clicking the links or copy them by using the button below.
        </div>

        <v-data-table :headers="headers" :items="missedPeptides" :items-per-page="10">
            <template v-slot:item.action="{ item }">
                <tooltip message="BLAST this peptide in an external window.">
                    <v-icon @click="openBlastUrl(item)">mdi-open-in-new</v-icon>
                </tooltip>
            </template>
        </v-data-table>

        <div
            style="margin-top: 8px; margin-bottom: 8px; display: flex; justify-content: center;"
            class="copy-button-container">
            <v-btn @click="copyToClipboard" color="primary">
                <v-icon left>
                    mdi-clipboard-text-outline
                </v-icon>
                Copy all to clipboard
            </v-btn>
        </div>
    </div>
</template>

<script lang="ts">
import Vue from "vue";
import { Component, Prop, Watch } from "vue-property-decorator";
import * as clipboard from "clipboard-polyfill";
import Utils from "./../../custom/Utils";
import Tooltip from "./../../custom/Tooltip.vue";
import { Peptide } from "@/business/ontology/raw/Peptide";

@Component({
    components: {
        Tooltip
    }
})
export default class MissingPeptidesList extends Vue {
    @Prop({ required: true })
    private missedPeptides: Peptide[] = [];

    private loading: boolean = true;

    private headers = [
        {
            text: "Peptide",
            align: "left",
            sortable: true,
            width: "80%",
            value: "value"
        },
        {
            text: "Actions",
            align: "center",
            sortable: false,
            width: "20%",
            value: "action"
        }
    ]

    private openBlastUrl(peptide: string) {
        const url: string = "http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&SET_SAVED_SEARCH=on" +
            "&USER_FORMAT_DEFAULTS=on&PAGE=Proteins&PROGRAM=blastp&QUERY=" + peptide + "&GAPCOSTS=11%201" +
            "&EQ_MENU=Enter%20organism%20name%20or%20id--completions%20will%20be%20suggested&DATABASE=nr" +
            "&BLAST_PROGRAMS=blastp&MAX_NUM_SEQ=100&SHORT_QUERY_ADJUST=on&EXPECT=10&WORD_SIZE=3" +
            "&MATRIX_NAME=BLOSUM62&COMPOSITION_BASED_STATISTICS=2&SHOW_OVERVIEW=on&SHOW_LINKOUT=on" +
            "&ALIGNMENT_VIEW=Pairwise&MASK_CHAR=2&MASK_COLOR=1&GET_SEQUENCE=on&NEW_VIEW=on&NUM_OVERVIEW=100" +
            "&DESCRIPTIONS=100&ALIGNMENTS=100&FORMAT_OBJECT=Alignment&FORMAT_TYPE=HTML&OLD_BLAST=false"
        Utils.openInBrowser(url);
    }

    private copyToClipboard() {
        clipboard.writeText(this.missedPeptides.join("\n"));
    }
}
</script>

<style>

</style>
