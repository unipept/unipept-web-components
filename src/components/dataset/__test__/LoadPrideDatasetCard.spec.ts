import { shallowMount, createLocalVue, mount } from "@vue/test-utils";
import MissingPeptidesList from "./../MissingPeptidesList.vue";
import Vue from "vue";
import Vuetify from "vuetify";
import Vuex from "vuex";
import Mock from "@/test/Mock";
import Assay from "@/logic/data-management/assay/Assay";
import flushPromises from "flush-promises";
import Utils from "./../../custom/Utils";
import LoadPrideDatasetCard from "./../LoadPrideDatasetCard.vue";


Vue.use(Vuetify);
Vue.use(Vuex);

const localVue = createLocalVue();

describe("LoadPrideDatasetCard", () => {
    let vuetify;
    let getters;
    let store;

    beforeEach(() => {
        vuetify = new Vuetify();

        store = new Vuex.Store({});
    });

    it("correctly fires create-assay event", (done) => {
        const wrapper = mount(LoadPrideDatasetCard, {
            store,
            localVue,
            vuetify
        });

        let assayIdInput = wrapper.find(".assay-id-input");
        assayIdInput.setValue("8500");

        let fetchPrideButton = wrapper.find(".fetch-pride-button");
        fetchPrideButton.trigger("click");

        expect(wrapper.find(".assay-id-input")).toMatchSnapshot();

        done();
    });
})
