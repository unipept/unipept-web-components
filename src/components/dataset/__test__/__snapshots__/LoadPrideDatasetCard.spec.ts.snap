// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadPrideDatasetCard correctly fires create-assay event 1`] = `
"<div class=\\"v-card v-card--flat v-sheet theme--light\\"><div class=\\"v-card__text\\"><h3>Load data from the PRIDE archive</h3> <p>You can easily load data from the <a href=\\"http://www.ebi.ac.uk/pride/\\" target=\\"_blank\\">PRIDE</a> data repository. Simply enter an assay id (e.g. 8500) in the field below and click the 'Load PRIDE Dataset' button. The corresponding dataset will then be fetched using the PRIDE API and loaded into the search form on the left.</p> <form novalidate=\\"novalidate\\" class=\\"v-form\\"><div class=\\"v-input theme--light v-text-field v-text-field--placeholder\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><div class=\\"v-text-field__slot\\"><label for=\\"input-7\\" class=\\"v-label v-label--active theme--light\\" style=\\"left: 0px; position: absolute;\\">Assay id</label><input id=\\"input-7\\" placeholder=\\"e.g. 8500\\" type=\\"text\\"></div><div class=\\"v-input__append-inner\\"><div class=\\"v-input__icon v-input__icon--\\"><i role=\\"button\\" class=\\"v-icon notranslate v-icon--link material-icons theme--light\\"></i></div></div></div><div class=\\"v-text-field__details\\"><div class=\\"v-messages theme--light\\"><span name=\\"message-transition\\" tag=\\"div\\" class=\\"v-messages__wrapper\\"></span></div></div></div></div></form> <div class=\\"card-actions\\"><button type=\\"button\\" class=\\"v-btn v-btn--contained theme--light v-size--default\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate v-icon--left mdi mdi-cloud-download theme--light\\"></i>
                Fetch PRIDE dataset
            </span></button> <!----></div> <form novalidate=\\"novalidate\\" class=\\"v-form\\"><div class=\\"v-input v-textarea theme--light v-text-field\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><div class=\\"v-text-field__slot\\"><label for=\\"input-15\\" class=\\"v-label theme--light\\" style=\\"left: 0px; position: absolute;\\">Peptide list</label><textarea name=\\"qs\\" spellcheck=\\"false\\" id=\\"input-15\\" rows=\\"7\\"></textarea></div></div><div class=\\"v-text-field__details\\"><div class=\\"v-messages theme--light\\"><span name=\\"message-transition\\" tag=\\"div\\" class=\\"v-messages__wrapper\\"></span></div></div></div></div> <span class=\\"v-tooltip v-tooltip--top\\"><div class=\\"v-tooltip__content\\" style=\\"left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;\\"></div><div class=\\"v-input theme--light v-text-field v-text-field--placeholder\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><div class=\\"v-text-field__slot\\"><label for=\\"input-20\\" class=\\"v-label v-label--active theme--light\\" style=\\"left: 0px; position: absolute;\\">Name this dataset</label><input name=\\"search_name\\" id=\\"input-20\\" placeholder=\\"e.g. Sample B5\\" type=\\"text\\"></div><div class=\\"v-input__append-inner\\"><div class=\\"v-input__icon v-input__icon--\\"><i role=\\"button\\" class=\\"v-icon notranslate v-icon--link material-icons theme--light\\"></i></div></div></div><div class=\\"v-text-field__details\\"><div class=\\"v-messages theme--light\\"><span name=\\"message-transition\\" tag=\\"div\\" class=\\"v-messages__wrapper\\"></span></div></div></div></div></span> <span class=\\"v-tooltip v-tooltip--top\\"><div class=\\"v-tooltip__content\\" style=\\"left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;\\"></div><div class=\\"v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox accent--text\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><div class=\\"v-input--selection-controls__input\\"><input aria-checked=\\"true\\" id=\\"input-26\\" role=\\"checkbox\\" type=\\"checkbox\\" value=\\"\\"><div class=\\"v-input--selection-controls__ripple accent--text\\"></div><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-checkbox-marked theme--light accent--text\\"></i></div><label for=\\"input-26\\" class=\\"v-label theme--light\\" style=\\"left: 0px; position: relative;\\"><span>Store dataset in browser's local storage</span></label></div></div></div></span></form> <div class=\\"card-actions\\"><button type=\\"button\\" class=\\"v-btn v-btn--contained theme--light v-size--default\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate v-icon--left mdi mdi-plus theme--light\\"></i>
                Add to selected datasets
            </span></button></div> <div class=\\"vue-snackbar-container\\"><!----></div></div></div>"
`;
